File: 07-progetto-fullstack/01-fast-food-fullstack/frontend+angular/src/app/service/prodotto.service.ts
// src/app/prodotto.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Prodotto {
  id: number;
  nome: string;
  categoria: string;
  prezzo: number;
  immagine: string;
}

@Injectable({
  providedIn: 'root'
})
export class ProdottoService {

  private baseUrl = 'http://localhost:8080/api/prodotti';

  constructor(private http: HttpClient) { }

  getAllProdotti(): Observable<Prodotto[]> {
    return this.http.get<Prodotto[]>(this.baseUrl);
  }

  getProdottiByCategoria(categoria: string): Observable<Prodotto[]> {
    return this.http.get<Prodotto[]>(`${this.baseUrl}/categoria/${categoria}`);
  }

  getPrezzoMedio(): Observable<number> {
  return this.http.get<number>(`${this.baseUrl}/prezzo-medio`);
}

getProdottoMenoCaro(): Observable<Prodotto> {
  return this.http.get<Prodotto>(`${this.baseUrl}/meno-caro`);
}

getProdottoPiuCaro(): Observable<Prodotto> {
  return this.http.get<Prodotto>(`${this.baseUrl}/piu-caro`);
}
}

File: 07-progetto-fullstack/01-fast-food-fullstack/backend+springboot/src/main/java/com/fastfood/service/ProdottoService.java
package com.fastfood.service;

import java.util.Comparator;
import java.util.List;

import org.springframework.stereotype.Service;

import com.fastfood.model.Prodotto;
import com.fastfood.repository.ProdottoRepository;

@Service
public class ProdottoService {
	private final ProdottoRepository repository;

	public ProdottoService(ProdottoRepository repository) {
		this.repository = repository;
	}

	public List<Prodotto> getAllProdotti() {
		return repository.findAll();
	}

	public List<Prodotto> getProdottiByCategoria(String categoria) {
		return repository.findByCategoria(categoria);
	}

	// Prezzo Medio
	public Double getPrezzoMedio() {
		return repository.findAll().stream().mapToDouble(Prodotto::getPrezzo).average().orElse(0.0);
	}

	// Prodotto meno caro
	public Prodotto getProdottoMenoCaro() {
		return repository.findAll().stream().min(Comparator.comparingDouble(Prodotto::getPrezzo)).orElse(null);
	}

	// Prodotto PIU caro
	public Prodotto getProdottoPiuCaro() {
		return repository.findAll().stream().max(Comparator.comparingDouble(Prodotto::getPrezzo)).orElse(null);
	}

}
